src/orchestrator.ts

```diff
@@ -3,106 +3,361 @@ import * as debug from "debug";
+                return val.EventType === EventType.TimerFired
+                    && val.TimerId === createdTimer.EventId;
+            })[0]
+            : undefined;
     }
 }
 
 // tslint:disable-next-line:max-classes-per-file
-export class Action {
+class TaskSet {
     constructor(
-        public type: string,
-        public name: string,
-        public input: any,
+        public isCompleted: boolean,
+        public actions: Action[],
+        public result?: any,
     ) { }
 }
 
 // tslint:disable-next-line:max-classes-per-file
-export class StateItem {
+class Task {
     constructor(
-        public name: string,
-        public input: any,
+        public isCompleted: boolean,
+        public action: Action,
+        public result?: any,
+        public timestamp?: Date,
+        public id?: number,
     ) { }
 }
+
+// tslint:disable-next-line:max-classes-per-file
+class TimerTask extends Task {
+    constructor(
+        public isCompleted: boolean,
+        public isCanceled: boolean,
+        public action: CreateTimerAction,
+        public result?: any,
+        public timestamp?: Date,
+        public id?: number,
+    ) { super(isCompleted, action, result, timestamp, id); }
+    public cancel() {
+        if (!this.isCompleted) {
+            this.action.isCanceled = true;
+        } else {
+            throw "Cannot cancel a completed task.";
+        }
+    }
+}
+
+// tslint:disable-next-line:max-classes-per-file
+enum ActionType {
+    CallActivity = 0,
+    CallActivityWithRetry = 1,
+    CallSubOrchestrator = 2,
+    CallSubOrchestratorWithRetry = 3,
+    ContinueAsNew = 4,
+    CreateTimer = 5,
+    WaitForExternalEvent = 6,
+}
+
+// tslint:disable-next-line:max-classes-per-file
+class HistoryEvent {
+    constructor(
+        public EventId: number,
+        public EventType: EventType,
+        public FireAt: Date,                // TimerScheduled, TimerFired
+        public IsPlayed: boolean = false,
+        public Input: any,                  // EventRaised
+        public Name: string,
+        public Result: string,
+        public TaskScheduledId: number,     // TaskCompleted
+        public TimerId: number,             // TimerFired
+        public Timestamp: Date,
+        public IsProcessed: boolean = false // internal use
+    ) { }
+}
+
+// tslint:disable-next-line:max-classes-per-file
+enum EventType {
+    ExecutionStarted = 0,
+    ExecutionCompleted = 1,
+    ExecutionFailed = 2,
+    ExecutionTerminated = 3,
+    TaskScheduled = 4,
+    TaskCompleted = 5,
+    TaskFailed = 6,
+    SubOrchestrationInstanceCreated = 7,
+    SubOrchestrationInstanceCompleted = 8,
+    SubOrchestrationInstanceFailed = 9,
+    TimerCreated = 10,
+    TimerFired = 11,
+    OrchestratorStarted = 12,
+    OrchestratorCompleted = 13,
+    EventRaised = 14,
+    ContinueAsNew = 15,
+    GenericEvent = 16,
+    HistoryState = 17,
+}
+
+// tslint:disable-next-line:max-classes-per-file
+abstract class Action {
```

Comment:
Abstract class is fine here, but you can also consider using an interface to enforce a public type exiting on an object. If you're not share common functionality, that might be more flexible.

Reply:
✔